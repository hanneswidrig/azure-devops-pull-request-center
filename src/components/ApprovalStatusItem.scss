.asi-container {
  padding: 5px 0;
  display: flex;
}
