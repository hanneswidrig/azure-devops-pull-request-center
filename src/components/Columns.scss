@import "azure-devops-ui/Core/_platformCommon.scss";

.work-button > span:last-child {
  font-size: 12px;
  line-height: 20px;
}

td.bolt-table-cell-content.no-padding.bolt-table-cell.bolt-list-cell > div {
  justify-content: center;
}

.invisible {
  opacity: 0;
  position: absolute;
}

.tooltip-overflow-parent {
  background-color: $backgroundColor;
  color: $backgroundColor;
  padding: 16px;
  border-radius: 4px;
}

.tooltip-overflow-child + .tooltip-overflow-child {
  margin-top: 8px;
}

.tooltip-overflow .bolt-tooltip-content {
  margin: 0;
  padding: 0;
  background-color: transparent;
  box-shadow: 0 1.6px 3.6px 0 rgba(0, 0, 0, 0.132), 0 0.3px 0.9px 0 rgba(0, 0, 0, 0.108);
  box-shadow: 0 1.6px 3.6px 0 var(--callout-shadow-color, rgba(0, 0, 0, 0.132)),
    0 0.3px 0.9px 0 var(--callout-shadow-secondary-color, rgba(0, 0, 0, 0.108));
}
