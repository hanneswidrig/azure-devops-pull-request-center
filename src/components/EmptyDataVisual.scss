@import 'node_modules/azure-devops-ui/Core/_platformCommon.scss';

.empty-data-container {
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  background-color: $backgroundColor;
  color: $primary-text;
}

.empty-data-text {
  font-size: 1.25rem;
  text-align: center;
  margin: 4rem 0;
}
