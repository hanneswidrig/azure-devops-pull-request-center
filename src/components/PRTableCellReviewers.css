.reviewers-container {
  display: flex;
  flex-wrap: wrap;
}

.reviewers-container-hidden {
  position: absolute;
  left: 1000%;
  opacity: 0;
}

td.bolt-table-cell-content.no-padding.bolt-table-cell.bolt-list-cell > div {
  justify-content: center;
}

.tooltip-overflow-container {
  background-color: var(--palette-neutral-80);
  color: var(--background-color);
  padding: 12px 12px 12px 8px;
  border-radius: 4px;
}

.tooltip-overflow-child:first-of-type {
  margin-top: 8px;
}

.tooltip-overflow-child + .tooltip-overflow-child {
  margin-top: 8px;
}

.bolt-pill-group.overflow-wrap > .bolt-pill-group-inner {
  margin-bottom: 8px;
}

.tooltip-overflow .bolt-tooltip-content {
  margin: 0;
  padding: 0;
  background-color: transparent;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 13.2%), 0 0 1px 0 rgba(0, 0, 0, 10.8%);
}
