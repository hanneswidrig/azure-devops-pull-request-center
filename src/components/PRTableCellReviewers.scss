@import 'azure-devops-ui/Core/_platformCommon.scss';

.reviewers-container {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
}

.reviewers-container-hidden {
  position: absolute;
  left: 1000%;
  opacity: 0;
}

td.bolt-table-cell-content.no-padding.bolt-table-cell.bolt-list-cell > div {
  justify-content: center;
}

.tooltip-overflow-container {
  background-color: $tooltip-background-color;
  color: $backgroundColor;
  padding: 12px 12px 12px 8px;
  border-radius: 4px;
}

.tooltip-overflow-child:first-of-type {
  margin-top: 8px;
}

.tooltip-overflow-child + .tooltip-overflow-child {
  margin-top: 8px;
}

.bolt-pill-group.overflow-wrap > .bolt-pill-group-inner {
  margin-bottom: 8px;
}

.tooltip-overflow .bolt-tooltip-content {
  margin: 0;
  padding: 0;
  background-color: transparent;
  box-shadow: 0 1.6px 3.6px 0 rgba(0, 0, 0, 0.132), 0 0.3px 0.9px 0 rgba(0, 0, 0, 0.108);
  box-shadow: 0 1.6px 3.6px 0 var(--callout-shadow-color, rgba(0, 0, 0, 0.132)),
    0 0.3px 0.9px 0 var(--callout-shadow-secondary-color, rgba(0, 0, 0, 0.108));
}
