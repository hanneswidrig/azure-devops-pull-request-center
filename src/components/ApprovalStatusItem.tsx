import React from 'react';

import { ITableColumn } from 'azure-devops-ui/Table';
import { IListBoxItem } from 'azure-devops-ui/ListBox';

import './ApprovalStatusItem.scss';
import { getReviewerVoteIconStatus } from './StatusIcon';

export const ApprovalStatusItem = (
  rowIndex: number,
  columnIndex: number,
  tableColumn: ITableColumn<IListBoxItem<{}>>,
  tableItem: IListBoxItem<{}>,
) => {
  return (
    <td key={rowIndex} className="bolt-list-box-text bolt-list-box-text-multi-select asi-container">
      {getReviewerVoteIconStatus(tableItem.id)}
      <span className="margin-left-8">{tableItem.text}</span>
    </td>
  );
};
