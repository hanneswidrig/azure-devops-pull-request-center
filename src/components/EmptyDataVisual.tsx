import React from 'react';

import './EmptyDataVisual.scss';

export const EmptyDataVisual = () => {
  return (
    <div className="empty-data-container">
      <div className="empty-data-text">
        <p>No Pull Requests available for this view.</p>
      </div>
    </div>
  );
};
