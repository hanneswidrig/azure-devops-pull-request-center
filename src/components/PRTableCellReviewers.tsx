import React, { useState, useCallback, useEffect } from 'react';
import { IdentityRefWithVote } from 'azure-devops-extension-api/Git/Git';

import { Pill } from 'azure-devops-ui/Pill';
import { Tooltip } from 'azure-devops-ui/TooltipEx';
import { PillGroup } from 'azure-devops-ui/PillGroup';
import { VssPersona } from 'azure-devops-ui/VssPersona';

import './PRTableCellReviewers.scss';
import { getVoteDescription } from '../lib/utils';
import { reviewerVoteToIColorLight } from '../lib/colors';
import { getReviewerVoteIconStatus, NoVote } from './StatusIcon';

type Widths = {
  index: number;
  width: number;
  totalWidth: number;
  isOverflow: boolean;
  tagElement: React.ReactElement;
};

const ReviewerOverflow: React.FC<{ hiddenElements: Widths[] }> = ({ hiddenElements }: { hiddenElements: Widths[] }) => (
  <Tooltip
    className="tooltip-overflow"
    renderContent={() => (
      <div className="tooltip-overflow-container flex-column rhythm-horizontal-4">
        <div className="flex-row flex-center justify-center">
          <NoVote className="vote-status no-vote" />
          <span style={{ fontWeight: 'bold' }}>Reviewers</span>
        </div>
        <div className="flex-column flex-center justify-start">
          {hiddenElements.length > 0 && hiddenElements.map(ve => ve.tagElement)}
        </div>
      </div>
    )}
  >
    <div className="tooltip-overflow-child">
      <Pill variant={2} size={1}>
        <span style={{ fontWeight: 'bold' }}>+{hiddenElements.length}</span>
      </Pill>
    </div>
  </Tooltip>
);

export const reviewerPill = (reviewer: IdentityRefWithVote, i: number, ref?: (node: HTMLDivElement | null) => void) => (
  <Tooltip
    key={i}
    renderContent={() => (
      <div className="flex-row rhythm-horizontal-4">
        <div className="flex-column">
          <div className="flex-row flex-center justify-center">
            <VssPersona
              className="icon-margin"
              imageUrl={reviewer._links['avatar'].href}
              size={'small'}
              displayName={reviewer.displayName}
            />
            <span>{reviewer.displayName}</span>
          </div>
          <div className="flex-row flex-center justify-start margin-top-8">
            {getReviewerVoteIconStatus(reviewer.vote)}
            <span className="font-weight-semibold margin-left-4">{getVoteDescription(reviewer.vote)}</span>
          </div>
        </div>
      </div>
    )}
  >
    <div ref={ref} className="tooltip-overflow-child">
      <Pill key={reviewer.id} variant={2} color={reviewerVoteToIColorLight(reviewer.vote)} size={1}>
        <div className="flex-row rhythm-horizontal-8">
          {getReviewerVoteIconStatus(reviewer.vote)}
          <span>{reviewer.displayName}</span>
        </div>
      </Pill>
    </div>
  </Tooltip>
);

type Props = { reviewers: IdentityRefWithVote[] };
export const PRTableCellReviewers: React.FC<Props> = ({ reviewers }: Props) => {
  const [widths, setWidths] = useState<Widths[]>([]);
  const [visibleElements, setVisibleElements] = useState<Widths[]>([]);
  const [hiddenElements, setHiddenElements] = useState<Widths[]>([]);

  const measureTag = useCallback(
    (node: HTMLDivElement | null) => {
      setWidths(widths => {
        if (node !== null) {
          const index = widths.length;
          const elementWidth = Math.ceil(node?.getBoundingClientRect().width) ?? 0;
          const totalWidth = [...widths.map(w => w.width), elementWidth].reduce((prev, curr) => prev + curr, 0);

          return [
            ...widths,
            {
              index: index,
              width: elementWidth,
              totalWidth: totalWidth,
              isOverflow: totalWidth > 600,
              tagElement: reviewerPill(reviewers[index], index),
            },
          ];
        }

        return widths;
      });
    },
    [reviewers],
  );

  useEffect(() => {
    if (widths.length > 0) {
      setVisibleElements(widths.filter(w => !w.isOverflow));
      setHiddenElements(widths.filter(w => w.isOverflow));
      setWidths([]);
    }
  }, [widths]);

  // useEffect(() => {
  //   setVisibleElements([]);
  //   setHiddenElements([]);
  // }, [reviewers]);

  return (
    <div className="reviewers-container">
      <div className="reviewers-container-hidden">{reviewers.map((r, i) => reviewerPill(r, i, measureTag))}</div>
      <PillGroup overflow={1}>
        {visibleElements.length > 0 && visibleElements.map(ve => ve.tagElement)}
        {hiddenElements.length > 0 && <ReviewerOverflow hiddenElements={hiddenElements} />}
      </PillGroup>
    </div>
  );
};
