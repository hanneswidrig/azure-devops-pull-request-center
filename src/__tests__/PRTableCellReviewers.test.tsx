import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup } from '@testing-library/react';

import { IdentityRefWithVote } from 'azure-devops-extension-api/Git/Git';
import { PRTableCellReviewers } from '../components/PRTableCellReviewers';

afterEach(cleanup);

describe('<PRTableCellReviewers />', () => {
  const reviewers: IdentityRefWithVote[] = [];

  test('renders successfully column without any reviewers', () => {
    const { container } = render(<PRTableCellReviewers reviewers={reviewers} />);
    expect(container).toMatchSnapshot();
  });
});
