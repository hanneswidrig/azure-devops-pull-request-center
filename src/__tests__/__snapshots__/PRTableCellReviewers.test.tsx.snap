// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PRTableCellReviewers /> renders successfully column without any reviewers 1`] = `
<div>
  <div
    class="reviewers-container"
  >
    <div
      class="reviewers-container-hidden"
    />
    <div
      class="bolt-pill-group flex-row overflow-wrap"
    >
      <div
        class="bolt-pill-group-inner flex-row"
      />
    </div>
  </div>
</div>
`;
